google/cloud/spanner/mutations.h

```diff
@@ -0,0 +1,234 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_MUTATIONS_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_MUTATIONS_H_
+
+#include "google/cloud/spanner/internal/tuple_utils.h"
+#include "google/cloud/spanner/value.h"
+#include "google/cloud/spanner/version.h"
+#include <google/spanner/v1/mutation.pb.h>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+/**
+ * A wrapper for Cloud Spanner mutations.
+ *
+ * In addition to the Data Manipulation Language (DML) based APIs, Cloud Spanner
+ * supports the mutation API, where the application describes data modification
+ * using a data structure instead of a SQL statement.
+ *
+ * This class serves as a wrapper for all mutations types. Use the builders,
+ * such as `InsertMutationBuilder` or `UpdateMutationBuilder` to create objects
+ * of this class.
+ *
+ * @see https://cloud.google.com/spanner/docs/modify-mutation-api
+ *   for more information about the Cloud Spanner mutation API.
+ */
+class Mutation {
+ public:
+  /**
+   * Creates an empty mutation.
+   *
+   * @note Empty mutations are not usable with the Cloud Spanner mutation API.
+   */
+  Mutation() = default;
+
+  Mutation(Mutation&&) noexcept = default;
+  Mutation& operator=(Mutation&&) noexcept = default;
+  Mutation(Mutation const&) = default;
+  Mutation& operator=(Mutation const&) = default;
+
+  friend bool operator==(Mutation const& lhs, Mutation const& rhs);
+  friend bool operator!=(Mutation const& lhs, Mutation const& rhs) {
+    return !(lhs == rhs);
+  }
+
+  /// Convert the mutation to the underlying proto.
+  google::spanner::v1::Mutation as_proto() && { return std::move(m_); }
```

Comment:
Oooo, for classes that wrap protos, pretty much 1:1, I like this technique (defined on rvalue). :-) Thanks!